drivers/siemens-sitop.c

```diff
@@ -0,0 +1,314 @@
+ * actually serial models.
+ *
+ * The FTDI USB-to-serial converters in the USB-models are programmed with a non-standard
+ * Product ID (mine had 0403:e0e3), but can be used with the normal ftdi_sio driver:
+ * # modprobe ftdi_sio
+ * # echo 0403 e0e3 > /sys/bus/usb-serial/drivers/ftdi_sio/new_id
+ *
+ * This can also be automated via a udev rule:
+ * ACTION=="add", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="e0e3", \
+ *   RUN+="/sbin/modprobe ftdi_sio", \
+ *   RUN+="/bin/sh -c 'echo 0403 e0e3 > /sys/bus/usb-serial/drivers/ftdi_sio/new_id'"
+ * 
+ * Use the following udev rule to create a persistent device name, for example /dev/ttyUPS:
+ * SUBSYSTEM=="tty" ATTRS{idVendor}=="0403", ATTRS{idProduct}=="e0e3" SYMLINK+="ttyUPS"
+ */
+
+#include "main.h"
+#include "serial.h"
+#include "limits.h"
+
+#define DRIVER_NAME	"Siemens SITOP UPS500 series driver"
+#define DRIVER_VERSION	"0.01"
+
+#define RX_BUFFER_SIZE 100
+/* The number of delimiter characters between two strings from the UPS: */
+#define NR_DELIMITER_CHARS 2
+
+/* driver description structure */
+upsdrv_info_t upsdrv_info = {
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Matthijs H. ten Berge <m.tenberge@awl.nl>",
+	DRV_EXPERIMENTAL,
+	{ NULL }
+};
+
+/* The maximum number of consecutive polls in which the UPS does not provide any data: */
+static int max_polls_without_data;
+/* The current number: */
+static int nr_polls_without_data;
+/* receive buffer */
+static char rx_buffer[RX_BUFFER_SIZE];
+static int rx_count;
+
+static struct {
+	int battery_alarm;
+	int dc_input_low;
+	int on_battery;
+	int battery_above_85_percent;
+} current_ups_status;
+
+/* remove n bytes from the head of rx_buffer, shift the remaining bytes to the start */
+static void rm_buffer_head(int n) {
+	if (rx_count <= n) {
+		/* nothing left */
+		rx_count = 0;
+		return;
+	}
+	rx_count -= n;
+	memmove(rx_buffer, rx_buffer + n, rx_count);
+}
+
+/* parse incoming data from the UPS.
+ * return true if something new was received.
+ */
+static int check_for_new_data() {
+	int new_data_received = 0;
+	int done = 0;
+	int num_received;
+	
+	while (!done) {
+		/* Get new data from the serial port.
+		 * No extra delay, just get the chars that were already buffered.
+		 */
+		num_received = ser_get_buf(upsfd, rx_buffer + rx_count, RX_BUFFER_SIZE - rx_count, 0, 0);
+		if (num_received < 0) {
+			/* comm error */
+			ser_comm_fail("error %d while reading", num_received);
+			/* try to re-open the serial port: */
+			if (upsfd) {
+				ser_close(upsfd, device_path);
+				upsfd = 0;
+			}
+			upsfd = ser_open_nf(device_path);
+			ser_set_speed_nf(upsfd, device_path, B9600);
+			done = 1;
+		} else if (num_received == 0) {
+			/* no (more) new data */
+			done = 1;
+		} else {
+			rx_count += num_received;
+			
+			/* parse received input data: */
+			while (rx_count >= (5 + NR_DELIMITER_CHARS)) { /* all input data are strings of 5 characters plus 1 or more delimiter characters */
```

Comment:
Good point. The delimiter characters are not mentioned in the protocol description, so by not relying on it will make the driver more robust. I changed this in the next commit.